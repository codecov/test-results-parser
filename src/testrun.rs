use pyo3::prelude::*;

use serde::Serialize;

static FRAMEWORKS: &[(&'static str, &'static str)] = &[
    ("pytest", "Pytest"),
    ("vitest", "Vitest"),
    ("jest", "Jest"),
    ("phpunit", "PHPUnit"),
];

static EXTENSIONS: &[(&str, &str)] = &[(".py", "Pytest"), (".php", "PHPUnit")];

fn check_substring_before_word_boundary(string: &str, substring: &str) -> bool {
    if let Some((_, suffix)) = string.to_lowercase().split_once(substring) {
        return suffix
            .chars()
            .next()
            .map_or(true, |first_char| !first_char.is_alphanumeric());
    }
    false
}

pub fn check_testsuites_name(testsuites_name: &str) -> Option<&'static str> {
    FRAMEWORKS
        .iter()
        .filter_map(|(name, framework)| {
            check_substring_before_word_boundary(testsuites_name, name)
                .then_some(framework)
                .map(|framework| *framework)
        })
        .next()
}

// i can't seem to get  pyo3(from_item_all) to work when IntoPyObject is also being derived
#[derive(FromPyObject, IntoPyObject, Clone, Debug, Serialize, PartialEq)]
pub struct Testrun {
    #[pyo3(item)]
    pub name: String,
    #[pyo3(item)]
    pub classname: String,
    #[pyo3(item)]
    pub duration: Option<f64>,
    #[pyo3(item)]
    pub outcome: String,
    #[pyo3(item)]
    pub testsuite: String,
    #[pyo3(item)]
    pub failure_message: Option<String>,
    #[pyo3(item)]
    pub filename: Option<String>,
    #[pyo3(item)]
    pub build_url: Option<String>,
    #[pyo3(item)]
    pub computed_name: Option<String>,
}

impl Testrun {
    pub fn framework(&self) -> Option<&'static str> {
        for (name, framework) in FRAMEWORKS {
            if check_substring_before_word_boundary(&self.testsuite, name) {
                return Some(framework);
            }
        }

        for (extension, framework) in EXTENSIONS {
            if check_substring_before_word_boundary(&self.classname, extension)
                || check_substring_before_word_boundary(&self.name, extension)
            {
                return Some(framework);
            }

            if let Some(message) = &self.failure_message {
                if check_substring_before_word_boundary(message, extension) {
                    return Some(framework);
                }
            }

            if let Some(filename) = &self.filename {
                if check_substring_before_word_boundary(filename, extension) {
                    return Some(framework);
                }
            }
        }
        None
    }
}

#[derive(Clone, Debug, Serialize, IntoPyObject)]
pub struct ParsingInfo {
    pub framework: Option<&'static str>,
    pub testruns: Vec<Testrun>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_detect_framework_testsuites_name_no_match() {
        let f = check_testsuites_name("whatever");
        assert_eq!(f, None)
    }

    #[test]
    fn test_detect_framework_testsuites_name_match() {
        let f = check_testsuites_name("jest tests");
        assert_eq!(f, Some("Jest"))
    }

    #[test]
    fn test_detect_framework_testsuite_name() {
        let t = Testrun {
            classname: "".to_string(),
            name: "".to_string(),
            duration: None,
            outcome: "pass".to_string(),
            testsuite: "pytest".to_string(),
            failure_message: None,
            filename: None,
            build_url: None,
            computed_name: None,
        };
        assert_eq!(t.framework(), Some("Pytest"))
    }

    #[test]
    fn test_detect_framework_filenames() {
        let t = Testrun {
            classname: "".to_string(),
            name: "".to_string(),
            duration: None,
            outcome: "pass".to_string(),
            testsuite: "".to_string(),
            failure_message: None,
            filename: Some(".py".to_string()),
            build_url: None,
            computed_name: None,
        };
        assert_eq!(t.framework(), Some("Pytest"))
    }

    #[test]
    fn test_detect_framework_example_classname() {
        let t = Testrun {
            classname: ".py".to_string(),
            name: "".to_string(),
            duration: None,
            outcome: "pass".to_string(),
            testsuite: "".to_string(),
            failure_message: None,
            filename: None,
            build_url: None,
            computed_name: None,
        };
        assert_eq!(t.framework(), Some("Pytest"))
    }

    #[test]
    fn test_detect_framework_example_name() {
        let t = Testrun {
            classname: "".to_string(),
            name: ".py".to_string(),
            duration: None,
            outcome: "pass".to_string(),
            testsuite: "".to_string(),
            failure_message: None,
            filename: None,
            build_url: None,
            computed_name: None,
        };
        assert_eq!(t.framework(), Some("Pytest"))
    }

    #[test]
    fn test_detect_framework_failure_messages() {
        let t = Testrun {
            classname: "".to_string(),
            name: "".to_string(),
            duration: None,
            outcome: "pass".to_string(),
            testsuite: "".to_string(),
            failure_message: Some(".py".to_string()),
            filename: None,
            build_url: None,
            computed_name: None,
        };
        assert_eq!(t.framework(), Some("Pytest"))
    }

    #[test]
    fn test_detect_build_url() {
        let t = Testrun {
            classname: "".to_string(),
            name: "".to_string(),
            duration: None,
            outcome: "pass".to_string(),
            testsuite: "".to_string(),
            failure_message: Some(".py".to_string()),
            filename: None,
            build_url: Some("https://example.com/build_url".to_string()),
            computed_name: None,
        };
        assert_eq!(t.framework(), Some("Pytest"))
    }
}
