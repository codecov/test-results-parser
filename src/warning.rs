use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct WarningInfo {
    pub message: String,
    pub location: u64,
}

impl WarningInfo {
    pub fn new(message: String, location: u64) -> Self {
        Self { message, location }
    }
}
