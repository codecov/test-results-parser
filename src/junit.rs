use pyo3::exceptions::PyException;
use pyo3::prelude::*;

use std::collections::HashMap;

use quick_xml::events::attributes::Attributes;
use quick_xml::events::Event;
use quick_xml::reader::Reader;

use crate::testrun::{Outcome, Testrun};

// from https://gist.github.com/scott-codecov/311c174ecc7de87f7d7c50371c6ef927#file-cobertura-rs-L18-L31
fn attributes_map(attributes: Attributes) -> HashMap<String, String> {
    return attributes
        .filter_map(|a| {
            if let Ok(attr) = a {
                let bytes = attr.value.into_owned();
                let value = String::from_utf8(bytes).unwrap();
                let key = String::from_utf8(attr.key.local_name().as_ref().to_vec()).unwrap();
                Some((key, value))
            } else {
                None
            }
        })
        .collect::<HashMap<_, _>>();
}

fn populate(attr_hm: &HashMap<String, String>, testsuite: String) -> Testrun {
    let name = format!(
        "{}::{}",
        attr_hm.get("classname").unwrap(),
        attr_hm.get("name").unwrap()
    );

    let duration = attr_hm.get("time").unwrap().to_string().parse().unwrap();

    Testrun {
        name,
        duration,
        outcome: Outcome::Pass,
        testsuite: testsuite,
    }
}

#[pyfunction]
pub fn parse_junit_xml(file_bytes: Vec<u8>) -> PyResult<Vec<Testrun>> {
    let file_string = String::from_utf8_lossy(&file_bytes).into_owned();
    let mut reader = Reader::from_str(file_string.as_str());
    reader.trim_text(true);

    let mut list_of_test_runs = Vec::new();

    let mut buf = Vec::new();

    let mut saved_testrun: Option<Testrun> = None;

    let mut curr_testsuite = String::new();

    loop {
        match reader.read_event_into(&mut buf) {
            Err(e) => {
                break Err(PyException::new_err(format!(
                    "Error at position: {} {:?}",
                    reader.buffer_position(),
                    e
                )))
            }
            Ok(Event::Eof) => {
                break Ok(list_of_test_runs);
            }
            Ok(Event::Start(e)) => match e.name().as_ref() {
                b"testcase" => {
                    let attr_hm = attributes_map(e.attributes());
                    saved_testrun = Some(populate(&attr_hm, curr_testsuite.clone()));
                }
                b"skipped" => {
                    let mut testrun = saved_testrun.unwrap();
                    testrun.outcome = Outcome::Skip;
                    saved_testrun = Some(testrun);
                }
                b"error" => {
                    let mut testrun = saved_testrun.unwrap();
                    testrun.outcome = Outcome::Error;
                    saved_testrun = Some(testrun);
                }
                b"failure" => {
                    let mut testrun = saved_testrun.unwrap();
                    testrun.outcome = Outcome::Failure;
                    saved_testrun = Some(testrun);
                }
                b"testsuite" => {
                    let attr_hm = attributes_map(e.attributes());

                    curr_testsuite = attr_hm.get("name").unwrap().to_string();
                }
                _ => (),
            },
            Ok(Event::End(e)) => match e.name().as_ref() {
                b"testcase" => {
                    list_of_test_runs.push(saved_testrun.unwrap());
                    saved_testrun = None;
                }
                _ => (),
            },
            Ok(Event::Empty(e)) => match e.name().as_ref() {
                b"testcase" => {
                    let attr_hm = attributes_map(e.attributes());
                    list_of_test_runs.push(populate(&attr_hm, curr_testsuite.clone()));
                }
                _ => (),
            },
            Ok(Event::Text(_)) => (),

            // There are several other `Event`s we do not consider here
            _ => (),
        }
        buf.clear()
    }
}
