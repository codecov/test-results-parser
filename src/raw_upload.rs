use anyhow::Context;

use base64::prelude::*;
use pyo3::prelude::*;
use std::collections::HashSet;
use std::io::prelude::*;

use flate2::bufread::ZlibDecoder;

use quick_xml::reader::Reader;
use serde::Deserialize;

use crate::junit::{get_position_info, use_reader};
use crate::testrun::ParsingInfo;
use crate::warning::WarningInfo;

#[derive(Deserialize, Debug, Clone)]
struct TestResultFile {
    filename: String,
    data: String,
}
#[derive(Deserialize, Debug, Clone)]
struct RawTestResultUpload {
    #[serde(default)]
    network: Option<HashSet<String>>,
    test_results_files: Vec<TestResultFile>,
}

#[derive(Debug, Clone)]
struct ReadableFile {
    filename: String,
    data: Vec<u8>,
}

const LEGACY_FORMAT_PREFIX: &[u8] = b"# path=";
const LEGACY_FORMAT_SUFFIX: &[u8] = b"<<<<<< EOF";

fn serialize_to_legacy_format(readable_files: Vec<ReadableFile>) -> Vec<u8> {
    let mut res = Vec::new();
    for file in readable_files {
        res.extend_from_slice(LEGACY_FORMAT_PREFIX);
        res.extend_from_slice(file.filename.as_bytes());
        res.extend_from_slice(b"\n");
        res.extend_from_slice(&file.data);
        res.extend_from_slice(b"\n");
        res.extend_from_slice(LEGACY_FORMAT_SUFFIX);
        res.extend_from_slice(b"\n");
    }
    res
}

// the warnings should be ordered by location because they're pushed to the vec as we parse
// so we can guarantee that warning[x].location >= warning[x - 1].location
// implicitly tested by warnings-junit.xml
fn format_warnings(input: &[u8], warnings: Vec<WarningInfo>, filename: &str) -> Vec<String> {
    let mut offset = 0;
    let mut result = Vec::new();
    let mut line = 1;
    let mut col = 0;
    let mut input_iter = input.iter();
    for warning in warnings {
        for bytes in input_iter
            .by_ref()
            .take((warning.location - offset) as usize)
        {
            if *bytes == b'\n' {
                line += 1;
                col = 0;
            } else {
                col += 1;
            }
        }
        offset += warning.location;
        result.push(if warning.message.contains('\n') {
            format!("{}\n\nAt {}:{} in {}", warning.message, line, col, filename)
        } else {
            format!("{} at {}:{} in {}", warning.message, line, col, filename)
        });
    }
    result
}

#[pyfunction]
#[pyo3(signature = (raw_upload_bytes))]
pub fn parse_raw_upload(raw_upload_bytes: &[u8]) -> anyhow::Result<(Vec<ParsingInfo>, Vec<u8>)> {
    let upload: RawTestResultUpload =
        serde_json::from_slice(raw_upload_bytes).context("Error deserializing json")?;
    let network: Option<HashSet<String>> = upload.network;

    let mut results: Vec<ParsingInfo> = Vec::with_capacity(upload.test_results_files.len());
    let mut readable_files: Vec<ReadableFile> = Vec::with_capacity(upload.test_results_files.len());

    for file in upload.test_results_files {
        let decoded_file_bytes = BASE64_STANDARD
            .decode(file.data)
            .context("Error decoding base64")?;

        let mut decoder = ZlibDecoder::new(decoded_file_bytes.as_slice());

        let mut decompressed_file_bytes = Vec::new();
        decoder
            .read_to_end(&mut decompressed_file_bytes)
            .context("Error decompressing file")?;

        let mut reader = Reader::from_reader(decompressed_file_bytes.as_slice());
        reader.config_mut().trim_text(true);
        let (framework, testruns, warnings) = use_reader(&mut reader, network.as_ref())
            .with_context(|| {
                let pos_conversion = reader.buffer_position().try_into();
                match pos_conversion {
                    Ok(pos) => {
                        let (line, col) = get_position_info(&decompressed_file_bytes, pos);
                        format!(
                            "Error parsing JUnit XML in {} at {}:{}",
                            file.filename, line, col
                        )
                    }
                    Err(_) => format!("Error parsing JUnit XML in {}", file.filename),
                }
            })?;

        let warning_strings: Vec<String> =
            format_warnings(&decompressed_file_bytes, warnings, &file.filename);

        let parsing_info = ParsingInfo {
            framework,
            testruns,
            warnings: warning_strings,
        };
        results.push(parsing_info);

        let readable_file = ReadableFile {
            data: decompressed_file_bytes,
            filename: file.filename,
        };
        readable_files.push(readable_file);
    }

    let readable_file = serialize_to_legacy_format(readable_files);

    Ok((results, readable_file))
}

#[cfg(test)]
mod tests {
    use base64::engine::general_purpose::STANDARD as BASE64_STANDARD;
    use base64::Engine;
    use flate2::Compression;
    use std::io::Write;

    use super::*;
    use insta::{assert_yaml_snapshot, glob};

    fn file_into_bytes(filename: &str) -> Vec<u8> {
        let upload = std::fs::read(filename).unwrap();
        let mut encoder = flate2::write::ZlibEncoder::new(Vec::new(), Compression::default());
        encoder.write_all(&upload).unwrap();
        let compressed = encoder.finish().unwrap();
        let base64_data = BASE64_STANDARD.encode(compressed);
        let upload_json = format!(
            r#"{{"network": [], "test_results_files": [{{"filename": "{}", "format": "base64+compressed", "data": "{}"}}]}}"#,
            filename.split('/').next_back().unwrap(),
            base64_data,
        );
        upload_json.into()
    }

    #[test]
    fn test_parse_raw_upload_success() {
        glob!("../tests", "*.xml", |path| {
            let upload_json = file_into_bytes(path.to_str().unwrap());
            let result = parse_raw_upload(&upload_json);
            match result {
                Ok((results, _)) => assert_yaml_snapshot!(results),
                Err(e) => {
                    assert_yaml_snapshot!(e.to_string());
                }
            }
        });
    }
}
