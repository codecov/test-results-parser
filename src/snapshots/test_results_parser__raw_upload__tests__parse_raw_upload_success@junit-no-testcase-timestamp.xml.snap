---
source: src/raw_upload.rs
expression: parse_raw_upload(&upload_json).unwrap().0
input_file: tests/junit-no-testcase-timestamp.xml
---
- framework: Pytest
  testruns:
    - name: "test_junit[junit.xml--True]"
      classname: tests.test_parsers.TestParsers
      duration: 0.186
      outcome: failure
      testsuite: pytest
      failure_message: aaaaaaa
      filename: ~
      build_url: ~
      computed_name: "tests.test_parsers.TestParsers::test_junit[junit.xml--True]"
    - name: "test_junit[jest-junit.xml--False]"
      classname: tests.test_parsers.TestParsers
      duration: 0.186
      outcome: pass
      testsuite: pytest
      failure_message: ~
      filename: ~
      build_url: ~
      computed_name: "tests.test_parsers.TestParsers::test_junit[jest-junit.xml--False]"
